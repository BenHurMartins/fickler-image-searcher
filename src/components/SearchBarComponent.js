import React, {useState, useEffect} from 'react';
import {StyleSheet, View} from 'react-native';
import {SearchBar} from 'react-native-elements';
import {Colors} from '../styles';
import MaterialIcons from 'react-native-vector-icons/MaterialIcons';
import {Button, ViewTermsHistory} from '../components';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {useSelector, useDispatch} from 'react-redux';
import * as types from '../reducers/actionsTypes';
import {getImages} from '../services/api';

MaterialIcons.loadFont();

const SearchBarComponent = () => {
  const dispatch = useDispatch();
  const {termsHistory, searchTerm} = useSelector(state => state.SearchReducer);
  const [startSearch, setStartSearch] = useState(false);

  const searchImages = () => {
    // Always when a search starts here we will define the page as 1
    if (searchTerm.trim().length > 2) {
      setStartSearch(false);
      dispatch({type: types.SET_LOADING, payload: true});
      dispatch({type: types.RESET_PAGE});
      storeTerms(searchTerm.trim());
      getImages(searchTerm.trim(), 1)
        .then(images => {
          setStartSearch(false);
          dispatch({type: types.SET_LOADING, payload: false});
          dispatch({type: types.SET_IMAGES, payload: images});
        })
        .catch(err => {
          dispatch({type: types.SET_LOADING, payload: false});
        });
    }
  };

  const storeTerms = async newTerm => {
    //every time a new search is made a term will be added
    try {
      let newTermsHistory = termsHistory;
      // Filter the terms to not add a duplicate value
      const found = termsHistory.find(element => element === newTerm);
      if (!found) {
        newTermsHistory.push(newTerm);
        const jsonValue = JSON.stringify(newTermsHistory);
        dispatch({type: types.SET_TERMS, payload: newTermsHistory});
        await AsyncStorage.setItem('terms-history', jsonValue);
      }
    } catch (e) {
      console.log(e);
    }
  };

  setSearchTerm = term => {
    setStartSearch(true);
    dispatch({type: types.SET_SEARCH_TERM, payload: term});
  };

  return (
    <View style={styles.viewContainer}>
      <SearchBar
        containerStyle={styles.containerStyle}
        inputContainerStyle={styles.inputContainerStyle}
        value={searchTerm}
        onChangeText={setSearchTerm}
        placeholder={'Search for ...'}
        inputStyle={styles.inputStyle}
        onSubmitEditing={searchImages}
        onFocus={() => {
          setStartSearch(true);
        }}
        blurOnSubmit={true}
        testID="input"
      />
      {startSearch ? <ViewTermsHistory setSearchTerm={setSearchTerm} /> : null}
      <Button
        title={'Go'}
        onPress={searchImages}
        customStyles={{width: '15%'}}
      />
    </View>
  );
};

export default SearchBarComponent;

const styles = StyleSheet.create({
  containerStyle: {
    backgroundColor: Colors.BACKGROUND_COLOR,
    borderBottomColor: 'transparent',
    borderTopColor: 'transparent',
    width: '80%',
  },
  inputContainerStyle: {
    backgroundColor: Colors.BACKGROUND_COLOR,
    borderColor: Colors.BLACK,
    borderBottomColor: Colors.BLACK,
    borderWidth: 1,
    borderBottomWidth: 1,
  },
  inputStyle: {
    color: Colors.BLACK,
  },
  viewContainer: {
    zIndex: 2,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
});
