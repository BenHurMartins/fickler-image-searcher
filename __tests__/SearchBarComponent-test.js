import 'react-native';
import React from 'react';
import {render, act, fireEvent} from '@testing-library/react-native';
import SearchBarComponent from '../src/components/SearchBarComponent';
import renderer from 'react-test-renderer';

describe('SearchBarComponent component', () => {
  it('SearchBarComponent renders correctly', () => {
    renderer.create(<SearchBarComponent />);
  });
  it('SearchBarComponent snapshot test', () => {
    const tree = renderer.create(<SearchBarComponent />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('Should show history terms', () => {
    const {getByText, getByTestId} = render(<SearchBarComponent />);
    const input = getByTestId('input');
    fireEvent.changeText(input, 'test');
    act(() => {
      expect(getByText('test')).not.toBe(null);
    });
  });
});
