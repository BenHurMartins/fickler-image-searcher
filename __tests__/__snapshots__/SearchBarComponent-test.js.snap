// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBarComponent component SearchBarComponent snapshot test 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
      "zIndex": 2,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderBottomColor": "transparent",
        "borderBottomWidth": 1,
        "borderTopColor": "transparent",
        "borderTopWidth": 1,
        "padding": 8,
        "width": "80%",
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderBottomColor": "#000000",
            "borderBottomWidth": 1,
            "borderColor": "#000000",
            "borderRadius": 3,
            "borderWidth": 1,
            "flexDirection": "row",
            "minHeight": 30,
            "overflow": "hidden",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "#86939e",
                        "fontSize": 18,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="iconIcon"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
        <TextInput
          allowFontScaling={true}
          blurOnSubmit={true}
          editable={true}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Search for ..."
          placeholderTextColor="#86939e"
          platform="default"
          rejectResponderTermination={true}
          style={
            Object {
              "color": "#000000",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 10,
              "minHeight": 40,
            }
          }
          testID="input"
          underlineColorAndroid="transparent"
          value="test"
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text
                    allowFontScaling={false}
                    style={
                      Array [
                        Object {
                          "color": "#86939e",
                          "fontSize": 18,
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                        Object {
                          "fontFamily": "Material Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        Object {},
                      ]
                    }
                    testID="iconIcon"
                  >
                    
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#000000",
        "borderRadius": 5,
        "height": 50,
        "justifyContent": "center",
        "opacity": 1,
        "width": "15%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 30,
        }
      }
    >
      Go
    </Text>
  </View>
</View>
`;
